### PR TITLE
Undeprecate `@api.one`.

### PR DESCRIPTION
**Pull request by yajo**
Description of the issue/feature this PR addresses:

Quoting from https://github.com/OCA/pylint-odoo/issues/45#issue-163891444:

> Odoo recommend us not using api.one: #8527
> After odoo use api.one in v9: 8ac40ac
> A issue for this case was created: #12408
> 
> Odoo answer: odoo/odoo#12408 (comment)

Current behavior before PR: The decorator works fine, but the real problem is that people expect what they should not from it. v8-v10 uses it, and no deprecation warning is issued in the logs.

The same as you could find yourself using `@api.one` where you shuldn't, now with the deprecation warning you keep seeing methods written in `@api.multi` where they shouldn't, like:

```
@api.multi
def do_something(self):
    for s in self:
        s.field1 = s.field2.get_field1()
```

Desired behavior after PR is merged: Instead of deprecating it, we document it. Now it's crystal clear what the decorator is for and what not, what you could expect from it if you use it for unexpected purposes.

So, after merging this, then this statement https://github.com/odoo/odoo/pull/8527#issuecomment-147336152 should not be true anymore, and thus there will be no need to deprecate this, given it's a great tool when used correctly.
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa

### CODE REVIEW

File: openerp/api.py [hunk 1]
```diff
@@ -404,30 +404,79 @@ def old_api(self, cr, uid, ids, *args, **kwargs):
...
+            chaining operations. Consider this 2 methods::
+
+                # Bad
+                @api.one
+                def get_partner_list(self):
+                    return self.partner_id
+
+                # Good
+                @api.one
+                @api.returns("res.partner")
```
Comment (pedrobaeza) [inline openerp/api.py]: This is for compatibility with old API, so you shouldn't mention this in a context where it dissapears.

Comment (Yajo) [inline openerp/api.py]: Please see https://github.com/odoo/odoo/pull/13199/files/aa83d4212b28ad5beec9e2369fd11ec01c9b6578#diff-0289d45b19b9b36491135288fc24fa7dR465. It still has its use.

Comment (pedrobaeza) [inline openerp/api.py]: OK, it seems weird to me as the initial definition of the decorator. Let's see what Odoo says.

**Comment by dreispt** [comment]
Ping @rco-odoo @odony !
Some "official" statement about the future of `@api.one` would be very appreciated.
My personal opinion is that is does much more good than bad: 
The runtime errors caused by not having a Singleton when one was expected are much harder to detect than the ones caused by returning the wrong data structure...

**Comment by Yenthe666** [comment]
@dreispt `@api.multi` in combination with `self.ensure_one()` works fine to be sure about records or recordsets.

**Comment by Yajo** [comment]
Well, it works fine if you don't care about raising exceptions to the end user.

**Comment by dreispt** [comment]
Annoyingly, it demands demands two additional lines of code (the ensure_one and the for loop) and indentation level. `@api.one` is more idiomatic.

**Comment by lasley** [comment]
I would disagree that `api.multi` and `api.ensure_one()` is the replacement for `api.one`. Isn't it the combination of `api.multi` with an inner records loop? Where does the error for a non-singleton come into play?

**Comment by odony** [comment]
As I mentioned in #12408 I'm quite in favor of keeping `@api.one` with a proper documentation for its quirks, so I like the idea of this PR. However I'd like to leave it up to @rco-odoo (including possibly rewording the docstring).

@Yajo thank you for the PR and the recap :-)

**Comment by osvalr** [comment]
:+1:

File: openerp/api.py [hunk 2]
```diff
@@ -404,30 +404,79 @@ def old_api(self, cr, uid, ids, *args, **kwargs):
...
+                self.name = name
 
-        may be called in both record and traditional styles, like::
+        The method may be called in both record and traditional styles, like::
 
             # recs = model.browse(cr, uid, ids, context)
-            names = recs.method(args)
-
-            names = model.method(cr, uid, ids, args, context=context)
+            names = recs.rename(args)
```
Comment (sebalix) [inline openerp/api.py]: ``rename()`` does not return anything here, affecting ``names`` is useless I think

File: openerp/api.py [hunk 3]
```diff
@@ -404,30 +404,79 @@ def old_api(self, cr, uid, ids, *args, **kwargs):
...
 
-        .. deprecated:: 9.0
+            names = model.rename(cr, uid, ids, args, context=context)
```
Comment (sebalix) [inline openerp/api.py]: Same here.

**Comment by bealdav** [comment]
:+1:

**Comment by adrianmerrall** [comment]
+1

**Comment by xmo-odoo** [comment]
> The same as you could find yourself using @api.one where you shuldn't, now with the deprecation warning you keep seeing methods written in @api.multi where they shouldn't, like:
> 
> ```python
@api.multi
def do_something(self):
    for s in self:
        s.field1 = s.field2.get_field1()
```
This is very much a method which should *not* use api.one, given api.one will generate an extra completely unnecessary list full of None.

**Comment by pedrobaeza** [comment]
So... what's the final result? Do you deprecate it or not?

**Comment by Yajo** [comment]
It seems a design spread all across Odoo to work with lists, while it should care more about iterables. Not sure if that's for XMLRPC compatibility, but surely all would go faster using more generators and less lists. You can start with `@api.one` in v10 if you want. Returning a generator of `None` would result in almost no performance impact, while keeping compatibility with methods that just use `for x in result`.

Another option would be to improve `@api.returns` giving it the ability to get a class as its first parameter. So, decorating with `@api.returns(list)` would transform that generator into a list, for places where it is required.

**Comment by xmo-odoo** [comment]
> Not sure if that's for XMLRPC compatibility, but surely all would go faster using more generators and less lists.

XMLRPC &al must reify everything either way, best case scenario would just be more overhead generating the same end-result (because there's an intermediate generator before getting to the list).

>  You can start with @api.one in v10 if you want. Returning a generator of None would result in almost no performance impact

A generator also wouldn't do anything (at all) when the result isn't iterated/reified/used. So assuming this:
```python
@api.one
def do_something():
    # do a thing
```
called like this:
```python
rs.do_something()
```
the callback would never execute.

**Comment by Yajo** [comment]
> Even the PR makes no sense, the whole point of api.one was to implicitly map and thus return things, yet the one recommendation the PR documents is to not return things.

I can remove or change that part if you want.

> best case scenario would just be more overhead generating the same end-result (because there's an intermediate generator before getting to the list).

Then, if lists are most efficient in this case, the problem you mentioned in https://github.com/odoo/odoo/pull/13199#issuecomment-241663430 is not such problem. After all iterating inside the `@api.multi` method or outside the `@api.one` one is the same CPU time... just a little more RAM occupied in a little moment before the `@api.one` list gets garbage-collected. And in the worst case, it falls into the abuse warning. I could improve that too if you want.

Another option we have is to never return nothing from an `@api.one`. You can `assert result is None` inside the decorator, and force everyone to use `@api.multi` when returning anything. Thus, the one method you could deprecate is `@api.returns`, not `@api.one`.

**Comment by xmo-odoo** [comment]
> I can remove or change that part if you want.

As I've said repeatedly, what I want is for that decorator not to be used anymore, no amount of documentation will make it obvious to developers when it could sensibly do half a dozen different things and the average developer will just latch on what they hope/expect it does rather than what it actually does. Case in point (yet again):

> Another option we have is to never return nothing from an @api.one. 

You're suggesting "undeprecating" `@api.one` by completely changing its semantics and making it behave in ways it was never intended to.

**Comment by guewen** [comment]
`@api.one` is better for extensibility, as it allows to call `super` before or after each iteration, while a `for` loop inside a `@api.multi` method only allows to call `super` before or after the whole iteration. On the other hand, I agree that the return is confusing as well as the name `@api.one` which is not straightforward.

**Comment by Yajo** [comment]
> You're suggesting "undeprecating" @api.one by completely changing its semantics and making it behave in ways it was never intended to.

You are saying that its original purpose is broken, so yes, let's change it. After all I think most developers just want it to avoid those ugly `for record in self` loops, and don't care much about the returns. That should go for master instead, of course.

Another option is to make the assertion only when no `@api.returns` is present. Or add a `result_wrapper` argument, defaulting to `None`, and force anyone to use `@api.one(list)` if they want that old behavior with unexpected outputs, that now will be expected because you explicitly told the decorator you want a list with the results.

**Comment by lasley** [comment]
> don't care much about the returns

There are times when even the returns are desirable such as `name_get`, `name_search`, etc.

**Comment by dreispt** [comment]
@xmo-odoo I'm sorry to insist, but I do feel that the ORM is better with `@api.one` than without it.
The main reason is that without it we are forced to use the `for x in self:` model methods again an again. Every time I wrote it a voice in the my back of my head says "there ought to be a better way".

The argument for the deprecation is "developers misunderstand what it returns". This by itself is not a good argument. It returns a list of the results, period. 

A better argument could be "what the function returns is unexpected". But IMO the design choice is reasonable. Any other alternative, including "return None", are more surprising the returning a list.

I think the real problem about `@api.one` is when people think "this will run only once" (rather than "once per record"), and then think "I can return this value" (a boolean, an action dict, etc), and then everything breaks down.

The solution for this problem is to use a better name: I suggest `@api.foreach` instead of `@api.one`.
For the specific use case above, such as the one I just described, developer can use a `api.ensure_one()` statement. Doing that looks a bit odd though, and arguably it would be more consistent with the rest of the API for it to be used as a decorator: `@api.ensure_one()`, to signal "I expect this run run only once".

In summary:
- deprecate `@api.one`, but
- add `@api.foreach` doing exactly what `@api.one` does
- add `@api.ensure_one` to the API

**Comment by andreparames** [comment]
@guewen how so? You can call super for each record using multi as well, e.g.:

```python
@api.multi
def hello(self):
    for record in self:
        super(myclass, record).hello()
```

**Comment by dreispt** [comment]
Maybe "explicit is better than implicit".
RIP `@api.one`

**Comment by pedrobaeza** [comment]
I don't agree. There's no official conclusion yet, and I don't want to let it die.

**Comment by dreispt** [comment]
@pedrobaeza The feedback I have from the R&D team is that it will be deprecated, and the current deprecation warning is to be kept. But I'm reopening.

**Comment by pedrobaeza** [comment]
Is this feedback private or is the one here in the issue? In the issue I see two sides: one from @xmo-odoo who wants to deprecate it, and another from @odony who sees this as something practical and useful.

**Comment by dreispt** [comment]
Publicly on [Twitter](https://twitter.com/bouvyd/status/768793218377015296) and [here](https://github.com/odoo/odoo/pull/13199#issuecomment-241681318), not any private feedback.

**Comment by rco-odoo** [comment]
For those who are interested, here are my thoughts on the subjet. In one sentence: **I am in favor of the removal of the decorator `@api.one`**. This opinion of mine is based on several observations.
* The decorator was introduced during the early stages of the development of the new API. At that point, the model, null record, record and recordset where all different beasts, and the decorator was called `@api.record`. I later simplified the design and merged most of those concepts together. The API decorators really were converters between the different kinds of model instances. This is no longer the case.
* The *name* is misleading: it suggests that the invoked recordset should be a single record. A name like `@api.foreach` or even `@api.mapeach` would have been better.
* The *semantics* is misleading: the decorator does a `map`-like application. Invoking the method returns a list with the results of the application of the function on the individual records. This has been a source of programming errors.
* The *semantics* is not really useful: we almost never need such a list of results in Odoo. The only places where it is convenient are methods that loop on a recordset and return nothing relevant. Compute methods represent the largest usage of `@api.one`. And in those cases, it only shortens a `for record in self`, and suppresses one level of indentation. The gain is actually pretty small. And, as @dreispt wrote, maybe "explicit is better than implicit."

Regarding the pull request,
* improving the documentation will not help: it will not make the code less confusing or the decorator more useful;
* renaming it will not really help: it may be clear about what it does, but that does not make it more useful;
* changing its semantics will confuse everybody: think about maintaining a module for two different versions of Odoo!

**Comment by Yajo** [comment]
I think that if you see a decorator above a method, that's explicit enough...
Also, about maintentance of several versions... If you deprecate it, anyway we will have to change lots of addons, so not a valid point IMHO.
So, if the whole problem of `@api.one` is having a list of results... Why not simply add an `assert result is None` after its invocation and forbid returning stuff? That's a single line of code that would prevent the programming errors you want to avoid and would give developers a way to forget about loopings in code, and with no overhead on production environments (asserts get removed). Everyone happy. :blush:

**Comment by NL66278** [comment]
:-1:  In my opinion the @api.one decorator makes it more difficult to see what code does. So let us get rid of it as soon as possible.

**Comment by rco-odoo** [comment]
I close this request, based on the arguments I gave in my comment above.

**Comment by Yajo** [comment]
Thanks, at least its crystal clear now :wink:

**Comment by blaggacao** [comment]
@rco-odoo actually I personally find creating a `@api.foreach` out of deliberate consideration in the future, can be an elegant option for working in the realm of recordsets. Such method cannot return nor raise, that's quite clear, I guess (no one would think about such thing in a for loop, anyways, except for very conscious reasons).
>And in those cases, it only shortens a for record in self, and suppresses one level of indentation.

As we know, Odoo is not adhering to PEP8, entirely, but people who do (for the good reasons PEP8 has been designed), will feel the noise of repeated for-loops and will probably end up writing their own decorator.

I **don't** want to fuel discussion, I only want to fuel thought respecting any publisher's sovereign decision that might result.

Sidenote: There could be possibilities to `yield` something useful in such loop and capturing (in a list/iterable) and returning what's yielded for later processing, for example through a `@api.yielded.apply(self.method)` or even a `@api.yielded.raise_validation_errors()`

In this line of thought, I think this would be a forced onto transparency gain about the nature of recordsets for everyone.

Think:
```python
    @api.foreach
    @api.yielded.apply(self.raise_validation_errors_in_my_own_beautiful_formatting)
    def _do_something_funny(self):
        if self.condition_is_true:
            yield "Error Msg"
```
```python
    @api.multi
    def _do_something_funny(self):
        result = []
        for rec in self:
            if rec.condition_is_true:
                result.append("Error Msg")
        if result:
            raise ValidationError ("Some errors:\n{0}".format('\n'.join(result))
```

**Comment by lasley** [comment]
FYI for anyone that runs across this, `api_foreach` module has been proposed for merge in OCA/server-tools#615
